frame/assets/src/lib.rs

```diff
@@ -1438,149 +1429,505 @@ impl<T: Config> Pallet<T> {
+			id,
+			target,
+			amount,
+			keep_alive,
+			respect_frozen,
+			best_effort,
+			|actual, details| {
+				// Check admin rights.
+				if let Some(check_admin) = maybe_check_admin {
+					ensure!(&check_admin == &details.admin, Error::<T>::NoPermission);
 				}
-			)?;
 
-			d.supply = d.supply.saturating_sub(burned);
+				debug_assert!(details.supply >= actual, "checked in prep; qed");
+				details.supply = details.supply.saturating_sub(actual);
+
+				Ok(())
+			},
+		)?;
+		Self::deposit_event(Event::Burned(id, target.clone(), actual));
+		Ok(actual)
+	}
+
+	// Reduces balance on a best-effort basis.
+	//
+	// Returns an error (in which case nothing happened) or the amount by which the balance was
+	// reduced.
+	//
+	// LOW-LEVEL: Does not attempt to maintain supply or emit events.
+	fn decrease_balance(
+		id: T::AssetId,
+		target: &T::AccountId,
+		amount: T::Balance,
+		keep_alive: bool,
+		respect_frozen: RespectFrozen,
+		best_effort: bool,
+		check: impl FnOnce(
+			T::Balance,
+			&mut AssetDetails<T::Balance, T::AccountId, DepositBalanceOf<T>>,
+		) -> DispatchResult,
+	) -> Result<T::Balance, DispatchError> {
+		if amount.is_zero() { return Ok(amount) }
+
+		let (actual, melted) =
+			Self::prep_debit(id, target, amount, keep_alive, respect_frozen, best_effort)?;
+
+		Asset::<T>::try_mutate(id, |maybe_details| -> DispatchResult {
+			let details = maybe_details.as_mut().ok_or(Error::<T>::Unknown)?;
 
-			Self::deposit_event(Event::Burned(id, target, burned));
-			Ok(burned)
-		})
+			check(actual, details)?;
+
+			Account::<T>::try_mutate_exists(id, target, |maybe_account| -> DispatchResult {
+				let mut account = maybe_account.take().unwrap_or_default();
+				debug_assert!(account.balance >= actual, "checked in prep; qed");
+
+				// Make the debit.
+				account.balance = account.balance.saturating_sub(actual);
+				*maybe_account = if account.balance < details.min_balance {
+					debug_assert!(account.balance.is_zero(), "checked in prep; qed");
+					Self::dead_account(id, target, details, account.sufficient);
+					None
+				} else {
+					Some(account)
+				};
+				Ok(())
+			})?;
+
+			Ok(())
+		})?;
+
+		if let Some(arg) = melted {
+			T::Freezer::melted(id, target, arg)
+		}
+
+		Ok(actual)
 	}
 
+	/// Returns the actual amount transferred; will never be less than `amount`; might be slightly
+	/// more unless `burn_dust` is `true`.
+	///
+	/// Will fail if the amount transferred is so small that it cannot create the destination due
+	/// to minimum balance requirements.
 	fn do_transfer(
 		id: T::AssetId,
-		source: T::AccountId,
-		dest: T::AccountId,
+		source: &T::AccountId,
+		dest: &T::AccountId,
 		amount: T::Balance,
 		maybe_need_admin: Option<T::AccountId>,
 		keep_alive: bool,
-	) -> DispatchResult {
-		let mut source_account = Account::<T>::get(id, &source);
-		ensure!(!source_account.is_frozen, Error::<T>::Frozen);
+		respect_frozen: RespectFrozen,
+		best_effort: bool,
+		burn_dust: bool,
```

Comment:
I think it would make sense to have these items together in a struct like `AssetSettings`.

It seems you would even use a default implementation of this setting often.

Reply:
I don't think there is a sensible default implementation. Each of the flags `keep_alive`, `burn_dust` and `best_effort` are individually important.

Reply:
Added flag structs.